Source/EventFlow.ServiceProvider/Registrations/ServiceCollectionServiceRegistration.cs

```diff
@@ -0,0 +1,219 @@
+
+            RegisterDescriptor(descriptor, keepDefault);
+        }
+
+        public void RegisterIfNotRegistered<TService, TImplementation>(Lifetime lifetime = Lifetime.AlwaysUnique)
+            where TService : class where TImplementation : class, TService
+        {
+            Register<TService, TImplementation>(lifetime, true);
+        }
+
+        public void Decorate<TService>(Func<IResolverContext, TService, TService> factory)
+        {
+            _decoratorService.AddDecorator(factory);
+        }
+
+        public IRootResolver CreateResolver(bool validateRegistrations)
+        {
+            var serviceProvider = _collection.BuildServiceProvider();
+            var resolver = new ServiceProviderRootResolver(serviceProvider, _collection);
+
+            if (validateRegistrations) resolver.ValidateRegistrations();
+
+            var bootstrapper = resolver.Resolve<Bootstrapper>();
+            using (var a = AsyncHelper.Wait)
+            {
+                a.Run(bootstrapper.Start(CancellationToken.None));
+            }
+
+            return resolver;
+        }
+
+        private ResolverContext GetContext(IServiceProvider provider)
+        {
+            var resolver = new ServiceProviderResolver(provider, _collection);
+            return new ResolverContext(resolver);
+        }
+
+        private void RegisterFactory(Type serviceType, Lifetime lifetime,
+            bool keepDefault,
+            Func<IServiceProvider, object> implementationFactory)
+        {
+            var serviceLifetime = GetLifetime(lifetime);
+            var descriptor = new ServiceDescriptor(serviceType, implementationFactory, serviceLifetime);
+
+            RegisterDescriptor(descriptor, keepDefault);
+        }
+
+        private void RegisterDescriptor(ServiceDescriptor descriptor, bool keepDefault = false)
+        {
+            if (keepDefault)
+                _collection.TryAdd(descriptor);
+            else
+                _collection.Add(descriptor);
+        }
+
+        private ServiceLifetime GetLifetime(Lifetime lifetime)
+        {
+            switch (lifetime)
+            {
+                case Lifetime.AlwaysUnique:
+                    return ServiceLifetime.Transient;
+                case Lifetime.Singleton:
+                    return ServiceLifetime.Singleton;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(lifetime), lifetime, null);
+            }
+        }
+
+        /// <summary>
+        ///     Ensures that the <see cref="Bootstrapper" /> is run in an ASP.NET Core
+        ///     environment when EventFlow is configured into an existing ServiceCollection
+        ///     instance and <see cref="CreateResolver" /> is not used.
+        /// </summary>
+        // ReSharper disable once ClassNeverInstantiated.Local
+        private class HostedBootstrapper : IHostedService
```

Comment:
@frankebersoll If anyone simply pulls this NuGet into a .NET Core console application, they won't know why their code isn't working as there's nothing to start the hosted service.

If this is used in a ASP.NET Core application, then the order of which the `StartAsync` of all hosted services determine if its valid to start using EventFlow. It might be better to simply make the dependency  on ASP.NET Core and use `IApplicationLifetime.ApplicationStarted` instead. It seems its [invoked before any hosted services](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/hosted-services#ihostedservice-interface).

Reply:
Might be better to remove and move the dependency to the ASP.NET Core package instead